rts/motoko-rts/src/gc/generational.rs

```diff
@@ -0,0 +1,499 @@
+    }
+
+    unsafe fn mark_all_reachable(&mut self) {
+        while let Some((obj, tag)) = pop_mark_stack() {
+            self.mark_fields(obj as *mut Obj, tag)
+        }
+    }
+
+    unsafe fn mark_fields(&mut self, object: *mut Obj, object_tag: Tag) {
+        visit_pointer_fields(
+            self,
+            object,
+            object_tag,
+            self.heap.limits.base,
+            |gc, field_address| {
+                let field_value = *field_address;
+                gc.mark_object(field_value);
+
+                #[cfg(debug_assertions)]
+                gc.barrier_coverage_check(field_address);
+            },
+            |gc, slice_start, array| {
+                const SLICE_INCREMENT: u32 = 127;
+                debug_assert!(SLICE_INCREMENT >= TAG_ARRAY_SLICE_MIN);
+                if array.len() - slice_start > SLICE_INCREMENT {
+                    let new_start = slice_start + SLICE_INCREMENT;
+                    // push an entire (suffix) array slice
+                    push_mark_stack(gc.heap.mem, array as usize, new_start);
+                    new_start
+                } else {
+                    array.len()
+                }
+            },
+        );
+    }
+
+    #[cfg(debug_assertions)]
+    unsafe fn barrier_coverage_check(&self, field_address: *mut Value) {
+        if self.strategy == Strategy::Full
+            && (field_address as usize) >= self.heap.limits.base
+            && (field_address as usize) < self.heap.limits.last_free
+            && (*field_address).points_to_or_beyond(self.heap.limits.last_free)
+        {
+            assert!(REMEMBERED_SET
+                .as_ref()
+                .unwrap()
+                .contains(Value::from_raw(field_address as u32)));
+        }
+    }
+
+    unsafe fn mark_root_mutbox_fields(&mut self, mutbox: *mut MutBox) {
+        let field_address = &mut (*mutbox).field;
+        if pointer_to_dynamic_heap(field_address, self.heap.limits.base) {
+            self.mark_object(*field_address);
+        }
+    }
+
+    unsafe fn should_be_threaded(&self, object: *mut Obj) -> bool {
+        let address = object as usize;
+        match self.strategy {
+            Strategy::Young => address >= self.heap.limits.last_free,
+            Strategy::Full => address >= self.heap.limits.base,
+        }
+    }
+
+    unsafe fn compact_phase(&mut self) {
+        if self.is_compaction_beneficial() {
+            self.thread_initial_phase();
+            self.move_phase();
+        }
+    }
+
+    fn is_compaction_beneficial(&self) -> bool {
+        self.survival_rate() < 0.95
+    }
+
+    fn generation_size(&self) -> usize {
+        let base = match self.strategy {
+            Strategy::Young => self.heap.limits.last_free,
+            Strategy::Full => self.heap.limits.base,
+        };
+        self.heap.limits.free - base
+    }
+
+    fn survival_rate(&self) -> f64 {
+        self.marked_space as f64 / self.generation_size() as f64
```

Comment:
I think it should be safe because it is a floating point division, resulting in `f64::INF`. The function `is_compaction_beneficial()` would then evaluate to `false`. I will add a comment on that.